src/csharp/Grpc.Core/Interceptors/ServerInterceptor.cs

```diff
@@ -0,0 +1,150 @@
+        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
+        public virtual async Task ServerStreamingServerHandler<TRequest, TResponse>(TRequest request, IServerStreamWriter<TResponse> responseStream, ServerCallContext context, ServerStreamingServerMethod<TRequest, TResponse> next)
+            where TRequest : class
+            where TResponse : class
+        {
+            await next(request, responseStream, context).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Server-side handler for intercepting bidi streaming calls.
+        /// </summary>
+        /// <typeparam name="TRequest">Request message type for this method.</typeparam>
+        /// <typeparam name="TResponse">Response message type for this method.</typeparam>
+        public virtual async Task DuplexStreamingServerHandler<TRequest, TResponse>(IAsyncStreamReader<TRequest> requestStream, IServerStreamWriter<TResponse> responseStream, ServerCallContext context, DuplexStreamingServerMethod<TRequest, TResponse> next)
+            where TRequest : class
+            where TResponse : class
+        {
+            await next(requestStream, responseStream, context).ConfigureAwait(false);
+        }
+
+        private static class WrapUtil<TRequest, TResponse>
+            where TRequest : class
+            where TResponse : class
+        {
+            public static UnaryServerMethod<TRequest, TResponse> Unary(UnaryServerMethod<TRequest, TResponse> handler, ServerInterceptor interceptor)
+            {
+                return async (request, context) => await interceptor.UnaryServerHandler<TRequest, TResponse>(request, context, handler).ConfigureAwait(false);
+            }
+
+            public static ClientStreamingServerMethod<TRequest, TResponse> ClientStreaming(ClientStreamingServerMethod<TRequest, TResponse> handler, ServerInterceptor interceptor)
+            {
+                return async (request, context) => await interceptor.ClientStreamingServerHandler<TRequest, TResponse>(request, context, handler).ConfigureAwait(false);
+            }
+
+            public static ServerStreamingServerMethod<TRequest, TResponse> ServerStreaming(ServerStreamingServerMethod<TRequest, TResponse> handler, ServerInterceptor interceptor)
+            {
+                return async (request, response, context) => await interceptor.ServerStreamingServerHandler<TRequest, TResponse>(request, response, context, handler).ConfigureAwait(false);
+            }
+
+            public static DuplexStreamingServerMethod<TRequest, TResponse> DuplexStreaming(DuplexStreamingServerMethod<TRequest, TResponse> handler, ServerInterceptor interceptor)
+            {
+                return async (request, response, context) => await interceptor.DuplexStreamingServerHandler<TRequest, TResponse>(request, response, context, handler).ConfigureAwait(false);
+            }
+        }
+
+        private Delegate WrapDelegate(Delegate d)
```

Comment:
All this reflection looks inefficient to me. I haven't quite got my head round how it's all used at the moment, but it would be great to avoid it...

Reply:
This process only happens at the startup time of the service once, so I don't think the potential performance impact matters. Unfortunately, I don't know of any other way to avoid this dance, since `dynamic` or using C# binder to find the appropriate overload require depending on additional libraries. We need this to map a non-generic object back to the correct parameterized-type overload. If you have a better way to accomplish this in mind, definitely let me know please.